demo/src/features/search-utils/base-provider.ts

```diff
@@ -0,0 +1,301 @@
+    return fileName?.startsWith('.') ?? false
+  }
+
+  protected handleFileChanges = (e: {
+    rawAdded?: vscode.Uri[]
+    rawUpdated?: vscode.Uri[]
+    rawDeleted?: vscode.Uri[]
+  }): void => {
+    // Batch file changes for better performance
+    this.bufferChanges(e)
+
+    if (this.debounceTimer) {
+      clearTimeout(this.debounceTimer)
+    }
+
+    this.debounceTimer = setTimeout(() => {
+      this.processBufferedChanges()
+    }, this.config.debounceDelay)
+  }
+
+  protected bufferChanges(e: {
+    rawAdded?: vscode.Uri[]
+    rawUpdated?: vscode.Uri[]
+    rawDeleted?: vscode.Uri[]
+  }): void {
+    // Add all added files
+    e.rawAdded?.forEach((uri: vscode.Uri) => {
+      this.changeBuffer.push({ type: FileChangeType.ADDED, resource: uri })
+    })
+
+    // Add all updated files
+    e.rawUpdated?.forEach((uri: vscode.Uri) => {
+      this.changeBuffer.push({ type: FileChangeType.UPDATED, resource: uri })
+    })
+
+    // Add all deleted files
+    e.rawDeleted?.forEach((uri: vscode.Uri) => {
+      this.changeBuffer.push({ type: FileChangeType.DELETED, resource: uri })
+    })
+  }
+
+  protected processBufferedChanges(): void {
+    if (this.changeBuffer.length === 0) return
+
+    SearchLogger.debug(`Processing ${this.changeBuffer.length} file changes`)
+
+    for (const change of this.changeBuffer) {
+      try {
+        this.applyFileChange(change)
+      } catch (error) {
+        SearchLogger.error(
+          `Error processing file change for ${change.resource.fsPath}`,
+          error as Error
+        )
+      }
+    }
+
+    this.changeBuffer.length = 0 // Clear buffer
+    SearchLogger.debug(`File cache updated. Total files: ${this.cachedFiles.size}`)
+  }
+
+  protected applyFileChange(change: FileChange): void {
+    switch (change.type) {
+      case FileChangeType.ADDED:
+        // Only add if it matches our inclusion criteria
+        if (
+          !PatternMatcher.shouldExclude(change.resource.fsPath, this.config.excludePatterns) &&
+          (this.config.includeHiddenFiles || !this.isHiddenFile(change.resource))
+        ) {
+          this.cachedFiles.add(change.resource)
+        }
+        break
+
+      case FileChangeType.DELETED:
+        this.cachedFiles.delete(change.resource)
+        break
+
+      case FileChangeType.UPDATED:
+        // For updates, we could re-validate the file, but for now just ensure it's still cached
+        if (
+          this.cachedFiles.has(change.resource) &&
+          PatternMatcher.shouldExclude(change.resource.fsPath, this.config.excludePatterns)
+        ) {
+          this.cachedFiles.delete(change.resource)
+        }
+        break
+    }
+  }
+
+  protected isWithinSizeLimit(uri: vscode.Uri): Promise<boolean> {
+    return new Promise((resolve) => {
+      this.fileService
+        .resolve(uri)
+        .then((stats) => {
+          resolve((stats.size ?? 0) <= this.config.maxFileSize)
+        })
+        .catch(() => resolve(false))
+    })
```

Comment:
I'm confused why you didn't just declare the method async and await/try-catch `this.fileService.resolve` :thinking: